packages/state-transition/src/cache/epochCache.ts

```diff
@@ -572,27 +634,78 @@ export class EpochCache {
     // ```
     this.epoch = computeEpochAtSlot(state.slot);
     this.syncPeriod = computeSyncPeriodAtEpoch(this.epoch);
+    // ELECTRA Only: Add current cpState.validators.length
+    // Only keep validatorLength for epochs after finalized cpState.epoch
+    // eg. [100(epoch 1), 102(epoch 2)].push(104(epoch 3)), this.epoch = 3, finalized cp epoch = 1
+    // We keep the last (3 - 1) items = [102, 104]
+    if (currEpoch >= this.config.ELECTRA_FORK_EPOCH) {
+      this.historicalValidatorLengths = this.historicalValidatorLengths.push(state.validators.length);
+
+      // If number of validatorLengths we want to keep exceeds the current list size, it implies
+      // finalized checkpoint hasn't advanced, and no need to slice
+      const hasFinalizedCpAdvanced =
+        this.epoch - state.finalizedCheckpoint.epoch < this.historicalValidatorLengths.size;
+
+      if (hasFinalizedCpAdvanced) {
+        // We use finalized cp epoch - this.epoch which is a negative number to keep the last n entries and discard the rest
+        this.historicalValidatorLengths = this.historicalValidatorLengths.slice(
+          state.finalizedCheckpoint.epoch - this.epoch
+        );
+      }
+    }
   }
 
   beforeEpochTransition(): void {
     // Clone (copy) before being mutated in processEffectiveBalanceUpdates
-    // NOTE: Force to use Uint8Array.slice (copy) instead of Buffer.call (not copy)
-    this.effectiveBalanceIncrements = Uint8Array.prototype.slice.call(this.effectiveBalanceIncrements, 0);
+    // NOTE: Force to use Uint16Array.slice (copy) instead of Buffer.call (not copy)
+    this.effectiveBalanceIncrements = Uint16Array.prototype.slice.call(this.effectiveBalanceIncrements, 0);
   }
 
   /**
    * Return the beacon committee at slot for index.
    */
   getBeaconCommittee(slot: Slot, index: CommitteeIndex): Uint32Array {
+    return this.getBeaconCommittees(slot, [index]);
+  }
+
+  /**
+   * Return a single Uint32Array representing concatted committees of indices
```

Comment:
may be worth adding a TODO to avoid concatting here, this will get expensive if it is actually used frequently with multiple indices

Reply:
this could be done in the shuffling refactor #6938 or leave TODO there
cc @matthewkeil

Reply:
I look at adding this now